lib/LanguageServerRename/Model/Renamer/VariableRenamer.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+namespace Phpactor\Extension\LanguageServerRename\Model\Renamer;
+
+use Generator;
+use Microsoft\PhpParser\Node;
+use Microsoft\PhpParser\Node\Expression\Variable;
+use Microsoft\PhpParser\Node\Parameter;
+use Microsoft\PhpParser\Node\PropertyDeclaration;
+use Microsoft\PhpParser\Parser;
+use Phpactor\Extension\LanguageServerRename\Model\NodeUtils2;
+use Phpactor\Extension\LanguageServerRename\Model\RenameLocationGroup;
+use Phpactor\Extension\LanguageServerRename\Model\RenameLocationsProvider;
+use Phpactor\Extension\LanguageServerRename\Model\RenameResult;
+use Phpactor\Extension\LanguageServerRename\Model\Renamer2;
+use Phpactor\TextDocument\ByteOffset;
+use Phpactor\TextDocument\ByteOffsetRange;
+use Phpactor\TextDocument\Location;
+use Phpactor\TextDocument\TextDocument;
+use Phpactor\TextDocument\TextDocumentLocator;
+use Phpactor\TextDocument\TextDocumentUri;
+use Phpactor\TextDocument\TextEdit;
+use Phpactor\TextDocument\TextEdits;
+
+class VariableRenamer implements Renamer2
+{
+    /** @var Parser */
+    private $parser;
+    /** @var NodeUtils2 */
+    private $nodeUtils;
+    /** @var TextDocumentLocator */
+    private $locator;
+    /** @var Node */
+    private $preparedNode;
+    /** @var RenameLocationsProvider */
+    private $locationsProvider;
+
+    public function __construct(
+        Parser $parser,
+        NodeUtils2 $nodeUtils,
+        TextDocumentLocator $locator,
+        RenameLocationsProvider $locationsProvider
+    ) {
+        $this->parser = $parser;
+        $this->nodeUtils = $nodeUtils;
+        $this->locator = $locator;
+        $this->locationsProvider = $locationsProvider;
+    }
+
+    public function prepareRename(TextDocument $textDocument, ByteOffset $offset): ?ByteOffsetRange
+    {
+        $rootNode = $this->parser->parseSourceFile((string)$textDocument);
+        $node = $rootNode->getDescendantNodeAtPosition($offset->toInt());
+        
+        if (
+            (
+                $node instanceof Variable &&
+                null === $node->getFirstAncestor(PropertyDeclaration::class)
+            )
+            || $node instanceof Parameter
+        ) {
+            $this->preparedNode = $node;
+            [ $range ] = $this->nodeUtils->getNodeNameRanges($node);
+            return $range;
+        }
+        return null;
+    }
+    /**
+     * {@inheritDoc}
+     */
+    public function rename(TextDocument $textDocument, ByteOffset $offset, string $newName): Generator
+    {
+        [ $oldName ] = $this->nodeUtils->getNodeNameTexts($this->preparedNode);
```

Comment:
Avoid temporal coupling -- it's fine to have a method `resolveRenameNode` and repeat. This method will not work by default given the public API. Apart from the coupling - this also makes this class stateful.

Reply:
I hated to use the same code twice. Thought long about it and decided since the LSP API is coupled I might be as well.